File: Data Structures & Algorithms/sliding-window-maximum/submission-1.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        # init
        dq = deque()
        left = 0
        res = []
        for i in range(k-1):
            while dq and nums[dq[-1]] < nums[i]: dq.pop()
            dq.append(i)
        

        # body
        for right in range(k-1, len(nums)):
            # add right
            while dq and nums[dq[-1]] < nums[right]: dq.pop()
            dq.append(right)

            # get max
            while dq[0] < left:
                dq.popleft()
            res.append(nums[dq[0]])

            # update left
            left += 1
        
        return res

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-1.java
class Solution {
    public int findMin(int[] nums) {
        int n = nums.length;
        int left = 0, right = n - 1;
        int last = nums[n - 1];
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] <= last) {   // mid 落在右段（含断崖底）→ 往左找边界
                right = mid - 1;
            } else {                   // mid 在左段 → 往右
                left = mid + 1;
            }
        }
        return nums[left];   // left = 右段的第一个元素 = 最小值。无 +1
    }
}